snuba/datasets/storages/events.py

```diff
@@ -288,30 +288,30 @@ def get_promoted_tags() -> Mapping[str, Sequence[str]]:
         for col in get_promoted_columns()
     }
 
-
-storage = WritableTableStorage(
-    schemas=StorageSchemas(read_schema=schema, write_schema=schema),
-    table_writer=TableWriter(
-        write_schema=schema,
-        stream_loader=KafkaStreamLoader(
-            processor=EventsProcessor(promoted_tag_columns),
-            default_topic="events",
-            replacement_topic="event-replacements",
-            commit_log_topic="snuba-commit-log",
-        ),
-        replacer_processor=ErrorsReplacer(
+def get_storage() -> Storage:
+    return WritableTableStorage(
```

Comment:
Why can;t you just import the storage object in the factory function instead of calling get_storage() every time? This opens the door to people importing directly this function and getting more instances, while right now we have one instance of the storage across the system.

Reply:
I was going to make a similar, but slightly different, comment to this — is the `get_storage` function actually necessary? It seems as though `storages.factory` could just import the `storage` object directly from here.

Reply:
Right now it's not useful, this is so in future there is a way to pass an optional config that can override the default one defined for that storage

Reply:
I assume this is the same reason why you had initialization parameters in the datasets in PR #844.
I don't think that's a good idea and I think we should avoid it.
Right now everything in the dataset definition has no mutable state. The state of the dataset instance and everything referenced from there is constant and statically defined (either in code or settings). At the time of the instantiation of the dataset class, everything is graved in stone, it cannot change, and any subsequent instantiation in the same release will yield the same exact result.

This has a few useful consequences:
1) we do not really care whether our datasets are singletons or not. For "not care" I mean none needs to make any assumption in their code regarding this. No need for assumptions implies simpler and harder to break system. We strive to keep them singletons nevertheless just for consistency, since it would make things easier should we decide to change that. (there is an asterisk for some behaviors around kafka but that's tech debt).
2) the dataset definition, being static per release (deploy) means it is like if it was a config object. We could potentially turn all the dataset definition into a massive structure in settings and everything would still work with the same abstraction, so it is quite flexible.

Passing parameters into the application entry points (today only dataset, but soon storage), erases those two benefits. Suddenly this class becomes stateful. While the state is still constant it is not static anymore. It depend on instantiation, so which instance you are dealing with in the application and whether the dataset is a singleton suddenly matters. 
I am not saying we should never do that, but we should have a very very strong reason for doing so.

If I understand it correctly, the reason is to allow CLI parameters to override dataset fields. I don't think that is a strong reason in that, we should simply not allow that. In a world where we configure multiple clusters in the settings and the dataset picks the cluster it wants, overriding this parameter on the CLI seems a symptom of a broken behavior. Who does that should be changing the config instead.
I believe this is happening for historical reasons (before datasets) and because our settings structure is suboptimal, but we should fix that instead of increasing the complexity of the dataset infrastructure.

* about Kafka. It is true that we override some parameters of kafka consumers. First we should not, but in any case that happens after we received the config from the dataset and we do not make changes to the state of the dataset object.

Reply:
I am going to drop this now since it's not needed in the current implementation. But it means that none of querylog and customizable hosts can be shipped until we have fully implemented clusters which may be some time off, and I'm not sure that's ideal.

Reply:
Are you sure about that. Maybe I am missing some use cases, but it seems to me only the cleanup and optimize script get a clickhouse host through the command line.
It is true, we use in production and potentially on premise users may be overriding (not documented though), but neither of those scripts use a Reader, or a dataset though. SO I don't think this should block your work.
Or am I missing some cases ?

Reply:
Yeah, it's just cleanup and optimize. But you do specify a dataset to run the command on and we'll have to fetch the appropriate host for that dataset from the storage. But I think you're right, we don't need reader and in the case where a different host is provided we can probably skip this and initialize a new clickhouse client.